src/Auth0.OidcClient.Core/Tokens/IdTokenValidator.cs

```diff
@@ -0,0 +1,151 @@
+﻿using Microsoft.IdentityModel.Tokens;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IdentityModel.Tokens.Jwt;
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+
+namespace Auth0.OidcClient.Tokens
+{
+    /// <summary>
+    /// Perform validation of a JWT ID token in compliance with specified <see cref="IdTokenRequirements"/>.
+    /// </summary>
+    internal static class IdTokenValidator
+    {
+        /// <summary>
+        /// Assert that all the <see cref="IdTokenRequirements"/> are met by a JWT ID token for a given point in time.
+        /// </summary>
+        /// <param name="required"><see cref="IdTokenRequirements"/> that should be asserted.</param>
+        /// <param name="rawIDToken">Raw ID token to assert requirements against.</param>
+        /// <param name="pointInTime">Optional <see cref="DateTime"/> to act as "Now" in order to facilitate unit testing with static tokens.</param>
+        /// <param name="signatureVerifier">Optional <see cref="ISignatureVerifier"/> to perform signature verification and token extraction. If unspecified
+        /// <see cref="AsymmetricSignatureVerifier"/> is used against the <paramref name="required"/> Issuer.</param>
+        /// <exception cref="IdTokenValidationException">Exception thrown if <paramref name="rawIDToken"/> fails to
+        /// meet the requirements specified by <paramref name="required"/>.
+        /// </exception>
+        /// <returns><see cref="Task"/> that will complete when the token is validated.</returns>
+        internal static async Task AssertTokenMeetsRequirements(this IdTokenRequirements required, string rawIDToken, DateTime? pointInTime = null, ISignatureVerifier signatureVerifier = null)
+        {
+            if (string.IsNullOrWhiteSpace(rawIDToken))
+                throw new IdTokenValidationException("ID token is required but missing.");
+
+            var token = DecodeToken(rawIDToken);
+
+            // For now we want to support HS256 + ClientSecret as we just had a major release.
+            // TODO: In the next major (v4.0) we should remove this condition as well as Auth0ClientOptions.ClientSecret
+            if (token.SignatureAlgorithm != "HS256")
```

Comment:
I saw you included a class `NoSignatureVerifier` but I found no usage of it anywhere. What happens if the algorithm is other than HS256 and RS256? e.g. "none".

Reply:
My vote is not to have NoSignatureVerifier unless it's required for certain backwards compatibility reasons.

Reply:
The `NoSignatureVerifier` class is just part of the Unit Test project so it can perform the bulk of the tests without needing to do signature verification although it is still setup to require RS256 on those tests.

`NoSignatureVerifier` is not available to other developers as it doesn't ship in the nuget package at all.

There's no way of getting a token through that isn't either `HS256` or `RS256` without access to the internals of the nuget package.  The package is configured to only allow one specific project to do that - our unit test library.

Reply:
> There's no way of getting a token through that isn't either HS256 or RS256

@damieng What looks odd on this line is that you are checking if the algorithm is not HS256 (assume RS256) and in that case use a signature verifier. 

What would happen if the algorithm is HS256, then? Is the HS256 signature checked by the wrapped OidcClient library for you?

Reply:
The point of the HS256 stuff is to keep existing behavior for back compat until the next major.

Whether the OidcClient library validates the signature on a HS256 or not seems somewhat irrelevant - we just want it to keep doing whatever it's doing until 4.0.